definitions.h

```diff
@@ -55,171 +60,350 @@ struct Vertex
-            w = wid;
-            setupInternal();
-            zeroOut();
+            scale.ptrM[i * this->columns + i] = scaleArray[i];
         }
 
-        // Assignment constructor
-        Buffer2D& operator=(const Buffer2D & ib)
+        if (initTF == false)
         {
-            w = ib.width();
-            h = ib.height();
-            setupInternal();
-            for(int r = 0; r < h; r++)
+            for (int i = 0; i < (this->rows * this->columns); i++)
             {
-                for(int c = 0; c < w; c++)
-                {
-                    grid[r][c] = ib[r][c];
-                }
+                this->ptrM[i] = scale.ptrM[i];
             }
+            initTF = true;
         }
-
-        // Set each member to zero 
-        void zeroOut()
+        //multiply to combine.
+        else
+        {
+            *this = scale * (*this);
+        }
+    }
+
+    /*************************************************************************
+	* Add translation to the matrix.
+	************************************************************************/
+    void translation(double transX, double transY, double transZ)
+    {
+        Matrix translate(this->rows, this->columns);
+        //initialize translation matrix.
+        translate.ptrM[1 * this->columns - 1] = transX;
+        translate.ptrM[2 * this->columns - 1] = transY;
+        translate.ptrM[3 * this->columns - 1] = transZ;
+        for (int i = 0; i < this->rows; i++)
+        {
+            translate.ptrM[i * this->columns + i] = 1;
+        }
+        // if the matrix has nothing in it make it equal
+        if (initTF == false)
         {
-            for(int r = 0; r < h; r++)
+            for (int i = 0; i < (this->rows * this->columns); i++)
             {
-                for(int c = 0; c < w; c++)
-                {
-                    grid[r][c] = 0;
-                }
+                this->ptrM[i] = translate.ptrM[i];
             }
+            initTF = true;
+        }
+        //Multiply to combine.
+        else
+        {
+            *this = translate * (*this);
+        }
+    }
+
+    /*****************************************************************
+	* Equals operator for the matrix
+	****************************************************************/
+    Matrix &operator=(const Matrix &rhs)
+    {
+        this->rows = rhs.rows;
+        this->columns = rhs.columns;
+        this->initTF = rhs.initTF;
+        if (this->ptrM != NULL)
+        {
+            delete[] ptrM;
+        }
+        ptrM = new double[this->rows * this->columns];
+        for (int i = 0; i < (this->rows * this->columns); i++)
+        {
+            this->ptrM[i] = rhs.ptrM[i];
+        }
+        return *this;
+    }
+
+    /*************************************************************************
+	* Multiplication operator for the matrix
+	************************************************************************/
+    Matrix &operator*=(const Matrix &rhs)
```

Comment:
Your * operator looks great! If you exclusively used 4x4 matrices you wouldn't need to check if the number of columns equals the number of rows but I guess it's nice to have the option of creating matrices of varying sizes. Good job!